Remove inplace

KimMcAlpine opened the pull request:
Remove the `in_place` module from katsdpcal which no longer functions with newer versions of dask

Commit:
SPR1-1790 Allow cal to work with newer dask

The latest version of dask + distributed that allows katsdpcal to function properly is 2021.9.0. In more recent versions of dask the `store_inplace` functionality is broken.

`store_inplace` allowed the pipeline to write changes to the numpy arrays backing the dask arrays used in the pipeline processing. In order do away with this functionality this PR introduces the following changes

1.) Instead of `apply_inplace` or `store_inplace` the pipeline computes the outcome of the required computation and recreates a new dask array based on these computed outcomes

2.) The `store_inplace` functionality allowed the pipeline process to store its computed flags in the same flag buffer array that was also accessed by the flag writer process. Instituting the changes in point 1.) above meant that the flag buffer was no longer being updated to include the computed cal `l1` pipeline flags. The flag buffer array is now explicitly passed to the pipeline function in reduction.py and manually updated with the computed l1 flags.

Commit:
Remove the inplace module and tests

ludwigschwardt commented on katsdpcal/reduction.py:
```diff
@@ -62,6 +62,71 @@ def init_flagger(parameters, dump_period):
 
 
+def slot_slices_to_time_slices(slices):
+    """Convert a list of slot slices to time slices"""
```
This assumes that `slices` have unit strides, which is obviously true if it was generated by `slots_slices`.

It might be useful to add an example like the one in `slots_slices`:
```
Example
-------
>>> slot_slices_to_time_slices([slice(2, 5, None), slice(6, 9, None), slice(0, 2, None)])
[slice(0, 3, None), slice(3, 6, None), slice(6, 8, None)]
```

KimMcAlpine replied:
Good idea, I have added the example

ludwigschwardt commented on katsdpcal/reduction.py:
```diff
@@ -62,6 +62,71 @@ def init_flagger(parameters, dump_period):
+        start = start+step
+
+
+def slots_slices(slots):
```
I have a similar-ish [bit of code](https://github.com/ska-sa/katdal/blob/0840fd86ca4954168cacf4cb785eb00afef121b4/katdal/lazy_indexer.py#L60) in katdal but sufficiently different that it doesn't make sense to import :-)

ludwigschwardt commented on katsdpcal/scan.py:
```diff
@@ -1385,16 +1378,12 @@ def rfi(self, flagger, auto_ant=False, sensors=None):
-        inplace.rename(data['cross_pol_orig'].flags)
-        inplace.rename(data['auto_pol_orig'].flags)
+        cross, auto = da.compute(data['cross_pol_flags'], data['auto_pol_flags'])
+        scandata.orig.cross_pol.flags = da.from_array(cross)
```
What happens if you assign
```Python
scandata.orig.cross_pol.flags = data['cross_pol_flags']
```
and leave the `compute` for higher-level functions? Is it too messy then?

KimMcAlpine replied:
a) It makes everything a lot slower as a lot of the subsequent calculations e.g. bsol, gsol, summarize etc compute the flags as part of their own `compute` sequences so you end up recomputing the flags multiple times and make the compute graphs of each subsequent step more complex 
b) I have to compute the flags in any case to store them in the numpy flag_array so it makes sense to just do it once here.

ludwigschwardt commented on katsdpcal/scan.py:
```diff
@@ -1385,16 +1378,12 @@ def rfi(self, flagger, auto_ant=False, sensors=None):
+        scandata.orig.auto_pol.flags = da.from_array(auto)
+
         self.cross_ant.reset_chunked()
+        self.auto_ant.reset_chunked()
```
Was this just something that was missed in the past?

KimMcAlpine replied:
Yes, I think so

ludwigschwardt reviewed (commented):
I have some minor comments and requests. My last thoughts before getting sidetracked for 1.5 months 😅 was whether it would be feasible to work the `slots` and `flags` getting passed around into other structures (like `data` and `Scan`) and whether `store_flags` can move into `Scan.rfi` then. We can discuss 🙂

KimMcAlpine commented:
> I have some minor comments and requests. My last thoughts before getting sidetracked for 1.5 months 😅 was whether it would be feasible to work the `slots` and `flags` getting passed around into other structures (like `data` and `Scan`) and whether `store_flags` can move into `Scan.rfi` then. We can discuss 🙂

I have tried to do this by adding the `slots` and shared  `flag` to the `Scan` class. I think your are right,  this is a nicer solution and not too complicated to do.

Commit:
SPP1-39 Hide flag updates within Scan.rfi

Previously after computing the flags with `Scan.rfi` the pipeline was
explicitly storing the computed flags in a shared memory flag array with
`store_flags`.
As storing the flags is always desirable after computing them it
makes more sense to combine these actions into a single step under
`Scan.rfi`

ludwigschwardt reviewed (approved):
Looks good!

ludwigschwardt commented:
🎉
